File: src/main/java/com/attilapalf/exceptional/services/GcmMessageService.java
package com.attilapalf.exceptional.services;

import java.util.List;
import java.util.stream.Collectors;

import javax.annotation.PostConstruct;

import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

import com.attilapalf.exceptional.entities.DevicesEntity;
import com.attilapalf.exceptional.entities.ExceptionInstancesEntity;
import com.attilapalf.exceptional.entities.UsersEntity;
import com.attilapalf.exceptional.messages.notifications.ExceptionNotification;
import com.attilapalf.exceptional.messages.notifications.FriendNotification;

/**
 * Created by palfi on 2015-08-29.
 */
@Service
public class GcmMessageService {
    private final String URL = "https://android.googleapis.com/gcm/send";
    private final String API_KEY = "AIzaSyCSwgwKHOuqBozM-JhhKYp6xnwFKs8xJrU";
    private final String PROJECT_NUMBER = "947608408958";

    private RestTemplate restTemplate = new RestTemplate();
    private HttpHeaders httpHeaders;

    @PostConstruct
    public void initHttpHeaders( ) {
        httpHeaders = new HttpHeaders();
        httpHeaders.setContentType( MediaType.APPLICATION_JSON );
        httpHeaders.set( "Authorization", "key=" + API_KEY );
    }

    public void sendFriendNotification( UsersEntity newUser, List<UsersEntity> friends ) {
        friends.forEach( friend -> pushNewFriendNotification( friend, newUser ) );
    }

    public void sendExceptionNotification( UsersEntity receiver, UsersEntity sender, ExceptionInstancesEntity exception ) {
        List<String> receiverGcmIds = getGcmIds( receiver );
        ExceptionNotification notification = new ExceptionNotification( receiverGcmIds, exception, receiver.getPoints(),
                sender.getPoints() );
        HttpEntity<ExceptionNotification> gcmRequestData = new HttpEntity<>( notification, httpHeaders );
        try {
            String gcmResponse = restTemplate.postForObject( URL, gcmRequestData, String.class );
        } catch ( Exception e ) {
            e.printStackTrace();
        }
    }

    private void pushNewFriendNotification( UsersEntity receiver, UsersEntity newUser ) {
        List<String> receiverGcmIds = getGcmIds( receiver );
        FriendNotification notification = new FriendNotification(
                newUser.getFacebookId(),
                newUser.getFirstName().trim() + " " + newUser.getLastName().trim(),
                receiverGcmIds );
        HttpEntity<FriendNotification> gcmRequestData = new HttpEntity<>( notification, httpHeaders );
        try {
            String gcmResponse = restTemplate.postForObject( URL, gcmRequestData, String.class );
        } catch ( RestClientException e ) {
            e.printStackTrace();
        }
    }

    private List<String> getGcmIds( UsersEntity receiver ) {
        return receiver.getDevices().stream().map( DevicesEntity::getGcmId ).collect( Collectors.toList() );
    }
}

File: src/main/java/com/attilapalf/exceptional/services/GcmMessageService.kt
package com.attilapalf.exceptional.services

import com.attilapalf.exceptional.entities.ExceptionInstance
import com.attilapalf.exceptional.entities.User
import com.attilapalf.exceptional.messages.notifications.AnswerNotification
import com.attilapalf.exceptional.messages.notifications.ExceptionNotification
import com.attilapalf.exceptional.messages.notifications.FriendNotification
import com.attilapalfi.exceptional.model.Question
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import javax.annotation.PostConstruct

/**
 * Created by palfi on 2015-08-29.
 */

public interface GcmMessageService {
    fun sendFriendNotification(newUser: User, friends: List<User>)
    fun sendExceptionNotification(sender: User, receiver: User,
                                  exception: ExceptionInstance, question: Question)

    fun sendAnswerNotification(exception: ExceptionInstance)
}

@Service
public class GcmMessageServiceImpl : GcmMessageService {
    private val URL = "https://android.googleapis.com/gcm/send"

    private val API_KEY = "AIzaSyCSwgwKHOuqBozM-JhhKYp6xnwFKs8xJrU"
    private val PROJECT_NUMBER = "947608408958"
    private lateinit var httpHeaders: HttpHeaders

    @PostConstruct
    public fun initHttpHeaders() {
        httpHeaders = HttpHeaders()
        httpHeaders.contentType = MediaType.APPLICATION_JSON
        httpHeaders.set("Authorization", "key=" + API_KEY)
    }

    override public fun sendFriendNotification(newUser: User, friends: List<User>) {
        val restTemplate = RestTemplate()
        friends.forEach { friend -> pushNewFriendNotification(friend, newUser, restTemplate) }
    }

    override public fun sendExceptionNotification(sender: User, receiver: User,
                                                  exception: ExceptionInstance, question: Question) {
        val receiverGcmIds = getGcmIds(receiver)
        val notification = ExceptionNotification(receiverGcmIds, exception, receiver.points, sender.points, question)
        val gcmRequestData = HttpEntity(notification, httpHeaders)
        try {
            val gcmResponse = RestTemplate().postForObject(URL, gcmRequestData, String::class.java)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun sendAnswerNotification(exception: ExceptionInstance) {
        val notification = AnswerNotification(getGcmIds(exception.fromUser), exception)
        val gcmRequestData = HttpEntity(notification, httpHeaders)
        try {
            val gcmResponse = RestTemplate().postForObject(URL, gcmRequestData, String::class.java)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun pushNewFriendNotification(receiver: User, newUser: User, restTemplate: RestTemplate) {
        val receiverGcmIds = getGcmIds(receiver)
        val notification = FriendNotification(
                newUser.facebookId,
                newUser.firstName.trim() + " " + newUser.lastName.trim(),
                receiverGcmIds)
        val gcmRequestData = HttpEntity(notification, httpHeaders)
        try {
            val gcmResponse = restTemplate.postForObject(URL, gcmRequestData, String::class.java)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun getGcmIds(receiver: User): List<String> {
        return receiver.devices.map { it.gcmId }
    }
}
